app/models/transformation_mapping_item.rb

```diff
@@ -24,4 +31,36 @@ def destination_cluster
       errors.add(:destination, "EMS type of destination cluster or cloud tenant must be in: #{destination_types}")
     end
   end
+
+  def source_datastore
+    source_storage = source
+
+    cluster_storages = source.hosts # Get hosts using this source storage
+                             .collect(&:ems_cluster) # How many clusters does each host have
+                             .collect(&:storages).flatten # How many storages each host is mapped to that belong to the cluster
+
+    unless cluster_storages.include?(source_storage)
+      storage_types = VALID_SOURCE_DATASTORE_TYPES.join(', ')
+      errors.add(:source, "Source type must be in: #{storage_types}")
+    end
+  end
+
+  def destination_datastore
+    destionation_storage = destination
+
+    dst_storages = if destination.kind_of?(Storage) # red hat
+                     destination.hosts
+                                .collect(&:ems_cluster)
+                                .collect(&:storages).flatten
+
+                   elsif destination.kind_of?(CloudVolume) # Openstack
+                     destination.cloud_tenant.cloud_volumes
+
+                   end
+
+    unless dst_storages.include?(destionation_storage)
```

Comment:
This looks like  very similar logic as above,  including the extra queries to get .  I'm not sure what we're checking.

Reply:
got fixed by using new filtering mechanism